### PR TITLE
feat(agg index): initial impl refresh index

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Initial impl refresh index.

Closes #issue

### CODE REVIEW

File: src/query/sql/src/executor/physical_plan.rs [hunk 1]
```diff
@@ -72,6 +73,35 @@ impl TableScan {
...
+    pub name_mapping: BTreeMap<String, IndexType>,
+    pub source: Box<DataSourcePlan>,
+
+    /// Only used for display
+    pub table_index: IndexType,
+    pub stat_info: Option<PlanStatsInfo>,
+
+    pub internal_column: Option<BTreeMap<FieldIndex, InternalColumn>>,
+}
+
```
Comment (RinChanNOWWW) [inline src/query/sql/src/executor/physical_plan.rs:91]: The structure is totally the same as `TableScan`, if we can reuse `TableScan` but add a new field to indicate it is for aggregating index?

**Comment by BohuTANG** [comment]
Please add the summary about this PR :)

**Comment by BohuTANG** [comment]
From the PR codes, it looks like a new feature, is it ready to add logic test for this PR or next?

**Comment by ariesdevil** [comment]
> From the PR codes, it looks like a new feature, is it ready to add logic test for this PR or next?↳

The logic tests are not suite for refresh index, we need design new test for this feature, this pr will merge first so that the it not stuck the read side by @RinChanNOWWW 

I'll continue work on refresh index includes: partial refresh by limit, add new field to index meta, and so on.

**Comment by RinChanNOWWW** [comment]
Better to add some unit tests to mock the index writing.

File: src/query/pipeline/transforms/src/processors/transforms/transform_aggregating_index.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+use common_pipeline_core::processors::processor::ProcessorPtr;
+use common_pipeline_core::processors::Processor;
+
+pub struct TransformAggIndex {
+    input: Arc<InputPort>,
+    output: Arc<OutputPort>,
+
+    input_data: Option<DataBlock>,
+    output_data: Option<DataBlock>,
+}
```
Comment (RinChanNOWWW) [inline src/query/pipeline/transforms/src/processors/transforms/transform_aggregating_index.rs:32]: Is this transform used now?

File: src/query/service/src/interpreters/interpreter_index_refresh.rs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                    let mut builder1 = PhysicalPlanBuilder::new(metadata.clone(), self.ctx.clone());
+                    let mut builder2 = PhysicalPlanBuilder::new(metadata.clone(), self.ctx.clone());
+                    (
+                        builder1.build(&table_scan_sexpr).await?,
+                        builder2.build(s_expr.as_ref()).await?,
+                        bind_context.output_schema(),
+                        bind_context.columns.clone(),
+                    )
+                }
+                _ => unreachable!(),
```
Comment (zhang2014) [inline src/query/service/src/interpreters/interpreter_index_refresh.rs:76]:
```suggestion
                _ => Err(),

```

File: src/query/service/src/interpreters/interpreter_index_refresh.rs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+impl RefreshIndexInterpreter {
+    pub fn try_create(ctx: Arc<QueryContext>, plan: RefreshIndexPlan) -> Result<Self> {
+        Ok(RefreshIndexInterpreter { ctx, plan })
+    }
+}
+
+#[async_trait::async_trait]
+impl Interpreter for RefreshIndexInterpreter {
```
Comment (zhang2014) [inline src/query/service/src/interpreters/interpreter_index_refresh.rs:51]: What happen if cluster mode

Comment (ariesdevil) [reply src/query/service/src/interpreters/interpreter_index_refresh.rs:52]: We decided to force use local mode now.

File: src/query/storages/fuse/src/operations/agg_index_sink.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        let mut result = DataBlock::new(vec![], block.num_rows());
+        for index in self.projections.iter() {
+            result.add_column(block.get_by_offset(*index).clone());
+        }
+        self.blocks.push(result);
+    }
+}
+
+#[async_trait]
+impl Processor for AggIndexSink {
```
Comment (RinChanNOWWW) [inline src/query/storages/fuse/src/operations/agg_index_sink.rs:137]: It's easier to `Impl AsyncSink for AggIndexSink`.

File: src/query/ee/tests/it/aggregating_index/index_refresh.rs [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    .await?;
+
+    let blocks = collect_file_names(&block_path)?;
+    assert!(blocks.len() > indexes.len());
+
+    // Refresh Index again
+    refresh_index(fixture.ctx(), "index1").await?;
+
+    let indexes = collect_file_names(&agg_index_path)?;
+    assert_eq!(blocks, indexes);
```
Comment (RinChanNOWWW) [inline src/query/ee/tests/it/aggregating_index/index_refresh.rs:148]: Also needs to check if the content of the index block is right.

File: src/query/storages/fuse/src/operations/agg_index_sink.rs [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            block_location,
+            location_data: HashMap::new(),
+            blocks: vec![],
+        });
+
+        Ok(ProcessorPtr::create(sinker))
+    }
+
+    fn process_block(&mut self, block: &mut DataBlock) {
+        let col = block.get_by_offset(self.block_location.unwrap());
```
Comment (sundy-li) [inline src/query/storages/fuse/src/operations/agg_index_sink.rs:101]: you can downcast it into `ValueRef` of StringType

**Review by sundy-li — APPROVED** [review]

**Review by RinChanNOWWW — APPROVED** [review]

**Pull request merged**